systems/primitives/random_source.cc

```diff
@@ -76,7 +76,8 @@ namespace internal {
 template <typename T>
 RandomSourceT<T>::RandomSourceT(RandomDistribution distribution,
                                 int num_outputs, double sampling_interval_sec)
-    : LeafSystem<T>(SystemTypeTag<RandomSourceT>{}),
+    : LeafSystem<T>(SystemScalarConverter(SystemTypeTag<RandomSourceT>{},
+          SystemScalarConverter::GuaranteedSubtypePreservation::kDisabled)),
       distribution_(distribution),
```

Comment:
May I ask why we need this `kDisabled`? I followed the implementation in `Adder` https://github.com/RobotLocomotion/drake/blob/256301dd8488c01d2721906c1c9da517012fd39e/systems/primitives/adder.cc#L10-L11.

Reply:
If you undo this change and run the new unit test, you'll see an exception raised because we have subclassing in play here.  Anyone with a `RandomSource` in their diagram would experience an exception during scalar conversion.